src/irmin-pack/inode.ml

```diff
@@ -759,6 +759,117 @@ struct
+          let lexbuf = Lexing.from_string s in
+          Sexplib.Sexp.scan_sexp lexbuf
+
+        let parse_from_file f = Sexplib.Sexp.load_sexp f
+        let sexp_of_hash h = Sexplib.Sexp.(Atom (Fmt.str "%a" pp_hash h))
+        let sexp_of_lazy_hash h = Sexplib.Std.sexp_of_lazy_t sexp_of_hash h
+
+        let sexp_of_metadata m =
+          Sexplib.Sexp.Atom (Fmt.str "%a" (Irmin.Type.pp metadata_t) m)
+
+        (* A kind is either a Content (B for blob) or a Node (N) *)
+        let sexp_of_kind k =
+          let open Sexplib.Sexp in
+          match k with
+          | `Contents (h, m) ->
+              List [ Atom "B"; sexp_of_hash h; sexp_of_metadata m ]
+          | `Node h -> List [ Atom "N"; sexp_of_hash h ]
+
+        let sexp_of_step s =
+          Sexplib.Sexp.Atom (Fmt.str "%a" (Irmin.Type.pp step_t) s)
+
+        let rec list_of_entries a =
+          Array.fold_left
+            (fun acc e ->
+              match e with
+              | None -> acc
+              | Some p -> (
+                  match p.I.target with
+                  | None -> acc
+                  | Some t -> List.rev_append (list_of_t t) acc))
+            [] a
+
+        and list_of_tree t = list_of_entries t.I.entries
+
+        and list_of_stepmap vs =
+          let open Sexplib.Sexp in
+          StepMap.fold
+            (fun s k acc -> List [ sexp_of_step s; sexp_of_kind k ] :: acc)
+            vs []
+
+        and list_of_v v =
+          match v with
+          | I.Values vs -> list_of_stepmap vs
+          | Tree t -> list_of_tree t
+
+        and list_of_t t = list_of_v t.I.v
+
+        let sexp_of_t t =
+          Sexplib.Sexp.(
+            List
+              ([
+                 Atom (string_of_int Conf.entries);
+                 Atom (string_of_int Conf.stable_hash);
+                 sexp_of_lazy_hash t.v.I.hash;
+               ]
+              @ list_of_t t.v))
+
+        exception Misconstructed_Sexp of string
+        exception Wrong_Config of (int * int) * (int * int)
+
+        let ( ?>= ) v = match v with Ok v -> v | Error (`Msg e) -> failwith e
+
+        let step_value_list_of_sexp (s : sexp list) =
+          let rec aux acc = function
+            | [] -> acc
+            | Sexplib.Sexp.(List [ Atom step; List value ]) :: tl ->
+                let value =
+                  match value with
+                  | [ Atom "B"; Atom hash; Atom metadata ] ->
+                      let hash = ?>=(Irmin.Type.of_string hash_t hash) in
+                      let meta =
+                        ?>=(Irmin.Type.of_string metadata_t metadata)
+                      in
+                      `Contents (hash, meta)
+                  | [ Atom "N"; Atom hash ] ->
+                      let hash = ?>=(Irmin.Type.of_string hash_t hash) in
+                      `Node hash
+                  | _ -> raise Exit
+                in
+                aux ((?>=(Irmin.Type.of_string step_t step), value) :: acc) tl
+            | _ -> raise Exit
```

Comment:
Shouldn't you raise `Misconstructed_Sexp` instead?

Reply:
Yes, it's the plan, I just got lazy because I have a `sexp list` and not a `sexp` and wanted to create the MR ;-)